File: lib/src/main/java/flooferland/showbiz/lowlevel/IShowFormat.java
package flooferland.showbiz.lowlevel;

import flooferland.chirp.safety.Result;
import flooferland.showbiz.lowlevel.show.ShowData;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

/** Common functions between formats */
public abstract interface IShowFormat {
    /** Generally used for loading the data from disk */
    public Result<ShowData, String> readFromStream(InputStream stream) throws IOException;
    
    /** Generally used for saving the data to disk */
    public void writeToStream(ShowData format, OutputStream stream) throws Exception;
}

File: src/main/kotlin/com/flooferland/bizlib/IShowFormat.kt
package com.flooferland.bizlib

import java.io.InputStream
import java.io.OutputStream
import java.nio.file.Files
import java.nio.file.Path

interface IShowFormat {
    /** @throws java.io.IOException */
    fun readFile(path: Path): RawShowData = read(Files.newInputStream(path))

    /** @throws java.io.IOException */
    fun read(stream: InputStream): RawShowData

    /** @throws java.io.IOException */
    fun write(stream: OutputStream, data: RawShowData)
}